[Merged by Bors] - feat(analysis/calculus/implicit): Implicit function theorem

urkud opened the pull request:
Fixes #1849

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/2749.20Implicit.20function.20theorem).

urkud committed:
feat(linear_algebra/projection): projection to a subspace

urkud committed:
Use explicit args

urkud committed:
Docstring

urkud committed:
Merge branch 'master' into projection

urkud committed:
Add implicit function theorem

sgouezel commented on src/analysis/normed_space/complemented.lean:
```diff
@@ -99,3 +99,21 @@ begin
+namespace continuous_linear_map
+
+/-- If `f : E → F` is a surjective continuous linear map between two Banach spaces
+and `g : E → ker f` is a continuous linear projection onto `ker f`, then `x ↦ (g x, f x)`
```
```suggestion
and `g : E → ker f` is a continuous linear projection onto `ker f`, then `x ↦ (f x, g x)`
```

sgouezel commented on src/linear_algebra/projection.lean:
```diff
@@ -165,6 +172,26 @@ begin
 end
 
+/-- If `f : E → F` is a surjective linear map and `g : E → ker f` is a linear
+projection onto `ker f`, then `x ↦ (g x, f x)` defines a linear equivalence
```
```suggestion
projection onto `ker f`, then `x ↦ (f x, g x)` defines a linear equivalence
```

sgouezel reviewed (commented):
Still marked as WIP, so I haven't looked at it properly yet, just mentioning two typos.

urkud committed:
Var name

urkud committed:
ker-range-prod

urkud committed:
Implicit args, move

urkud committed:
Implicit - snapshot

urkud committed:
Implicit args

urkud committed:
Complemented - snapshot

urkud committed:
Merge branch 'master' into implicit-function-thm

urkud committed:
feat(linear_algebra/basic): range of `linear_map.prod`

Also make `ker_prod` a `simp` lemma.

urkud committed:
feat(linear_algebra/projection): add `equiv_prod_of_surjective_of_is_compl`

If kernels of two surjective linear maps `f`, `g` are complement subspaces,
then `x ↦ (f x, g x)` defines a  linear equivalence.

I also add a version of this equivalence for continuous maps.

urkud committed:
Docs

urkud committed:
Merge branch 'projection' into equiv-prod

urkud committed:
Merge branch 'master' into projection

urkud committed:
Merge branch 'projection' into equiv-prod

urkud committed:
Merge branch 'equiv-prod' into implicit-function-thm

urkud committed:
Add nolint

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+theorem.
+
+Then we define `implicit_function_of_complemented`: implicit function defined by `f (g z y) = z`,
+where `f : E → F` is a function strictly differentiable at `a` such that its defivative `f'`
```
```suggestion
where `f : E → F` is a function strictly differentiable at `a` such that its derivative `f'`
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+where `f : E → F` is a function strictly differentiable at `a` such that its defivative `f'`
+is surjective and has a `complemented` kernel.
+
+Finally, if the codomain of `f` is a finitely dimensional space, then we can automatically prove
```
```suggestion
Finally, if the codomain of `f` is a finite dimensional space, then we can automatically prove
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+
+/-- Implicit function theorem. If `f : E → F` and `g : E → G` are two maps strictly differentiable
+at `a`, their derivatives `f'`, `g'` are surjective, and the kernels of these derivatives are
+complement subspaces of `E`, then `x ↦ (f x, g x)` defines a local homeomorphism between
```
```suggestion
complementary subspaces of `E`, then `x ↦ (f x, g x)` defines a local homeomorphism between
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+/-- Implicit function theorem. If `f : E → F` and `g : E → G` are two maps strictly differentiable
+at `a`, their derivatives `f'`, `g'` are surjective, and the kernels of these derivatives are
+complement subspaces of `E`, then `x ↦ (f x, g x)` defines a local homeomorphism between
+`E` and `F × G`. In particular, `f x = f a` is locally homeomorphic to `G`. -/
```
```suggestion
`E` and `F × G`. In particular, `{x | f x = f a}` is locally homeomorphic to `G`. -/
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+
+/-- Implicit function theorem. If `f : E → F` and `g : E → G` are two maps strictly differentiable
+at `a`, their derivatives `f'`, `g'` are surjective, and the kernels of these derivatives are
+complement subspaces of `E`, then `implicit_function_of_is_compl_ker` is the unique (germ of a) map
```
```suggestion
complementary subspaces of `E`, then `implicit_function_of_is_compl_ker` is the unique (germ of a) map
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+  ∀ᶠ x in 𝓝 φ.pt, φ.implicit_function (φ.left_fun x) (φ.right_fun x) = x :=
+φ.has_strict_fderiv_at.eventually_left_inverse
+
+lemma to_implicit_function
```
The name is strange, since you're not in the `has_strict_fderiv_at` namespace. I think `has_strict_fderiv_at` should be part of the name.

urkud replied:
```suggestion
lemma implicit_function_has_strict_fderiv_at
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+
+variables (f f')
+
+/-- Data used to apply generic implicit function theorem to the case of a strictrly differentiable
```
```suggestion
/-- Data used to apply the generic implicit function theorem to the case of a strictly differentiable
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+variables (f f')
+
+/-- Data used to apply generic implicit function theorem to the case of a strictrly differentiable
+ map such that its derivative is surjective has has a complemented kernel. -/
```
```suggestion
 map such that its derivative is surjective and has a complemented kernel. -/
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+
+/-- Data used to apply generic implicit function theorem to the case of a strictrly differentiable
+ map such that its derivative is surjective has has a complemented kernel. -/
+@[simp] def implicit_function_data_of_complemented (hf : has_strict_fderiv_at f f' a)
```
`[simp]` or `[simps]`?

urkud replied:
`@[simps]` makes Lean reach an unreachable code.

gebner replied:
Issue please.

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+  is_compl_ker := linear_map.is_compl_of_proj (classical.some_spec hker) }
+
+/-- A local homeomorphism between E` and `F × f'.ker` sending level surfaces of `f`
+to horizontal subspaces. -/
```
Should it be vertical subspaces? Your model for `f` is the first projection, with kernel the vertical space `f.ker = f'.ker`.

urkud replied:
```suggestion
to vertical subspaces. -/
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,308 @@
+  (f a, (0 : f'.ker)) ∈ (hf.implicit_to_local_homeomorph f f' hf').target :=
+by apply mem_implicit_to_local_homeomorph_of_complemented_target
+
+/-- `implicit_function_of_complemented` sends `(z, y)` to a point in `f ⁻¹' z`. -/
```
```suggestion
/-- `implicit_function` sends `(z, y)` to a point in `f ⁻¹' z`. -/
```

sgouezel reviewed (commented):
I like the interface. This will need some specialization later on (over reals or complexes, strict differentiability follows from begin `C^1`, and then the implicit function is as smooth as the map), but this can be added later without any trouble thanks to the great flexibility of your interface, and the maximal generality it uses.

I still think docstrings could be expanded a little bit. In particular the docstring for `implicit_function_data` could explain what this means (and emphasize the fact that with this maximally general version one can choose the complementary subspace, while it will be chosen arbitrarily in the forthcoming API). And the file level docstring could give an example (say in two real variables, locally parameterizing the circle around `1`), first at the informal level, and then illustrating what the interface for the previous informal discussion would be like.

Did I say that this looks great? This looks great!

urkud committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

urkud committed:
Merge branch 'master' into equiv-prod

urkud committed:
Merge branch 'equiv-prod' into implicit-function-thm

urkud committed:
Fix compile

urkud committed:
Docs

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,389 @@
+(see `implicit_function_data.implicit_function`) such that for `(y, z)` close enough to `(f a, g a)`
+we have `f (φ y z) = y` and `g (φ y z) = z`.
+
+We also prove a formula for $$\frac{\partial\varphi}{\partial z}$$.
```
```suggestion
We also prove a formula for $$\frac{\partial\varphi}{\partial z}.$$
```

sgouezel commented on src/analysis/calculus/implicit.lean:
```diff
@@ -0,0 +1,389 @@
+  right_range := linear_map.range_eq_of_proj (classical.some_spec hker),
+  is_compl_ker := linear_map.is_compl_of_proj (classical.some_spec hker) }
+
+/-- A local homeomorphism between E` and `F × f'.ker` sending level surfaces of `f`
```
```suggestion
/-- A local homeomorphism between `E` and `F × f'.ker` sending level surfaces of `f`
```

sgouezel committed:
Apply suggestions from code review
